### ISSUE TITLE
Provide an Option to Use Path-Style-Access with S3 Snapshot Repositories

### ISSUE THREAD

**Issue by modularTaco** [crate/crate#11385]
**Use case**:
Some s3 storage providers only support path-style access and no vhost style access.

To support all s3 storage providers, I think it would be a good idea to add an option to `CREATE REPOSITORY` which can be used to force either path or vhost style.  
[ElasticSearch has added such an option](https://github.com/elastic/elasticsearch/pull/44046) 

According to the [Amazon Docs](https://docs.aws.amazon.com/AWSJavaSDK/latest/javadoc/com/amazonaws/services/s3/AmazonS3Builder.html#setPathStyleAccessEnabled-java.lang.Boolean-%5BAWSdocumentation%5D) the library will try to auto-detect if vhost or path style should be used.  
Therefore I've tried the following things to connect a path-style s3 storage provider as Snapshot repository:  
- Specify the bucket as part of `base_path` and set `bucket` to `''` -> the repository gets created, snapshots get created, but the snapshots cannot be listed (`select * from sys.snapshots` is always empty, even if the snapshots are present in the bucket)
- Use a bucketname which is not a valid hostname -> those bucket names were rejected by the s3 storage provider
- Use the IP-address instead of the hostname -> This works (the library now decides to use path-style) but comes with a disadvantage: i need to use http instead of https

**Comment by mfussenegger** [comment]
> Use the IP-address instead of the hostname -> This works (the library now decides to use path-style) but comes with a disadvantage: i need to use http instead of https

Doesn't it work if you set `protocol` to `https`?

Which s3 storage providers only support path-style access?

**Comment by modularTaco** [comment]
> > Use the IP-address instead of the hostname -> This works (the library now decides to use path-style) but comes with a disadvantage: i need to use http instead of https
> 
> Doesn't it work if you set `protocol` to `https`?
> 
> Which s3 storage providers only support path-style access?

no, setting the `protocol` to `https` and using the ip-adress does not work in my case as the certificate is only valid for the domain name and not for the ip adress

The s3 storage provider we use has added vhost support in the mean time as beta feature, so the issue is resolved for us.  
Nevertheless i think it's a good idea to give the choice to the user if vhost style or path style should be used, instead of relying on the automatic decision from the aws sdk

**Comment by mfussenegger** [comment]
Thanks for the clarification.

We currently don't have any concrete plans to work on it. If there are more users asking for it that could change - otherwise we'd be happy to accept contributions.

**Comment by mfussenegger** [comment]
In CrateDB 6.3, s3 uses path style by default.
